src/java.base/macosx/classes/sun/nio/fs/BsdFileSystem.java

```diff
@@ -71,13 +76,84 @@ public Set<String> supportedFileAttributeViews() {
         return SupportedFileFileAttributeViewsHolder.supportedFileAttributeViews;
     }
 
+    /**
+     * Clones the file whose path name is {@code src} to that whose path
+     * name is {@code dst} using the {@code clonefile} system call.
+     *
+     * @param src the path of the source file
+     * @param dst the path of the destination file (clone)
+     * @param followLinks whether to follow links
+     *
+     * @return 0 on success, or IOStatus.UNSUPPORTED_CASE if the call
+     *         does not work with the given parameters
+     */
+    private int clone(UnixPath src, UnixPath dst, boolean followLinks)
+        throws IOException
+    {
+        int flags = followLinks ? 0 : CLONE_NOFOLLOW;
+        try {
+            BsdNativeDispatcher.clonefile(src, dst, flags);
+        } catch (UnixException x) {
+            switch (x.errno()) {
+                case ENOTSUP: // cloning not supported by filesystem
+                case EXDEV:   // src and dst on different filesystems
+                case ENOTDIR: // problematic path parameter(s)
+                    return IOStatus.UNSUPPORTED_CASE;
+                default:
+                    x.rethrowAsIOException(src, dst);
+                    return IOStatus.THROWN;
+            }
+        }
+
+        return 0;
+    }
+
     @Override
     protected int directCopy(int dst, int src, long addressToPollForCancel)
         throws UnixException
     {
         return directCopy0(dst, src, addressToPollForCancel);
     }
 
+    @Override
+    protected void copyFile(UnixPath source,
+                            UnixFileAttributes attrs,
+                            UnixPath target,
+                            Flags flags,
+                            long addressToPollForCancel)
+        throws IOException
+    {
+        // Attempt to clone the source unless cancellation is not possible,
+        // or attributes are not to be copied
+        if (addressToPollForCancel == 0 && flags.copyPosixAttributes) {
```

Comment:
It wouldn't be wrong to the check for copyPosixAttributes here, in which case they would be copied by default (the spec allows for that). The only issue is that someone might become dependent on this and then be surprised when they code behaves differently when run on another platform.

I'm curious if there is a way to test how clone behaves when it fails. What you have is okay although the unlink might only be needed when chown fails.

Reply:
> The only issue is that someone might become dependent on this and then be surprised when they code behaves differently when run on another platform.

This is a good reason to leave it as-is.

> I'm curious if there is a way to test how clone behaves when it fails. What you have is okay although the unlink might only be needed when chown fails.

All the described error numbers except EIO are things which one would expect the call to verify before actually starting to clone, so for those cases it would be unexpected to find that the destination exists after the failure. Provoking an EIO failure might be difficult.